changefeedccl: modify changefeed job payload to support db-level changefeeds

KeithCh opened the pull request:
Update and populate changefeed job payload to 
support db-level changefeeds.

Epic: CRDB-1421
Resolves: #147370

Release note: None

KeithCh committed:
changefeedccl: rename TableID to DescID

To support changefeed database and schema targets,
it would be more accurate to use DescID instead of
TableID in the ChangefeedTargetSpecification and
Target structs.

Epic: CRDB-1421
Resolves: #147370

Release note: None

KeithCh commented:
> Second, what format will we need for DB level feeds with an include or exclude list?

It be a simple list, something like 
``` CREATE CHANGEFEED FOR DATABASE db with include_tables='a,b' ```

>  Will we ever have changefeed specs with both DB and table level target?

Our backend should be able to support this easily, it's just a matter of whether this is something customers want

asg0451 commented on pkg/jobs/jobspb/jobs.proto:
```diff
@@ -1099,7 +1099,7 @@ message ChangefeedTargetSpecification {
     COLUMN_FAMILY = 2;
 
     // Add TargetTypes for database, secondary index, etc. when implemented
```
alter this comment

asg0451 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -475,6 +475,19 @@ func evalCursor(
+	} else if changefeedStmt.Level == tree.ChangefeedLevelDatabase {
+		targetList.Databases = tree.NameList{tree.Name(changefeedStmt.DatabaseTarget)}
+	}
+	return targetList
```
nit: else { return errors.AssertionFailedf("unknown etc")

KeithCh replied:
I'll just do a if else case since there is currently no third value it can take on.

asg0451 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -899,31 +965,36 @@ func validateSettings(ctx context.Context, needsRangeFeed bool, execCfg *sql.Exe
 	initialHighWater hlc.Timestamp,
-) (map[tree.TablePattern]catalog.Descriptor, error) {
+) (
+	map[tree.TablePattern]catalog.Descriptor,
+	[]catalog.DatabaseDescriptor,
+	[]catalog.Descriptor,
```
can you add names for these pls &/ change the return values to make them make more sense

asg0451 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -542,65 +555,118 @@ func createChangefeedJobRecord(
+	_, hasMultiChangefeedPriv, err := checkPrivilegesForDescriptor(ctx, p, databaseDesc)
+	if err != nil {
+		return nil, err
+	}
+	tableToSchemaLookup := buildTableToSchemaLookup(allDescs)
+	// If the user has database-level or schema-level changefeed privileges,
+	// they can create table-level changefeeds on all tables in the database
+	// or schema (except if they don't have select privileges on the table and
+	// it is a core changefeed)
+	if changefeedStmt.Level == tree.ChangefeedLevelTable {
+		for _, desc := range targetTableDescs {
+			if table, isTable := desc.(catalog.TableDescriptor); isTable {
+				schemaDesc, ok := tableToSchemaLookup[table.GetID()]
+				if !ok {
+					return nil, errors.Errorf("expected to find a schema descriptor for table %s", table.GetName())
+				}
+				_, hasChangefeed, err := checkPrivilegesForDescriptor(ctx, p, schemaDesc)
+				if err != nil {
+					return nil, err
+				}
+				hasMultiChangefeedPriv = hasMultiChangefeedPriv && hasChangefeed
 
-			hasSelect, hasChangefeed, err := checkPrivilegesForDescriptor(ctx, p, desc)
-			if err != nil {
-				return nil, err
+				if err := changefeedvalidators.ValidateTable(specs, table, tolerances); err != nil {
+					return nil, err
+				}
+				for _, warning := range changefeedvalidators.WarningsForTable(table, tolerances) {
+					p.BufferClientNotice(ctx, pgnotice.Newf("%s", warning))
+				}
+
+				hasSelect, hasChangefeed, err := checkPrivilegesForDescriptor(ctx, p, desc)
+				if err != nil {
+					return nil, err
+				}
+				hasSelectPrivOnAllTables = hasSelectPrivOnAllTables && hasSelect
+				hasChangefeedPrivOnAllTables = hasMultiChangefeedPriv || (hasChangefeedPrivOnAllTables && hasChangefeed)
```
can you make these authz changes its own pr pls

asg0451 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -542,65 +555,118 @@ func createChangefeedJobRecord(
-
-	if err != nil {
-		return nil, err
+	databaseDesc := allDescs[0]
```
this is sus

asg0451 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -1989,3 +2075,30 @@ func maybeUpgradePreProductionReadyExpression(
+// buildTableToSchemaLookup processes the targetAndParentDescs list and creates a lookup
+// from table descriptor ID to schema descriptor. The targetAndParentDescs list is ordered
+// by database, schema1, table1_1,...,table1_n, schema2, table2_1,.., table2_m, etc.
+func buildTableToSchemaLookup(
```
lets not do it this way

asg0451 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -866,7 +932,7 @@ Few hours to a few days range are appropriate values for this option.`
 		Username:    p.User(),
 		DescriptorIDs: func() (sqlDescIDs []descpb.ID) {
-			for _, desc := range targetDescs {
+			for _, desc := range targetTableDescs {
```
remember to update this to include dbs too (in the next pr)

KeithCh committed:
changefeedccl: modify changefeed job payload to support db-level changefeeds

Update and populate changefeed job payload to
support db-level changefeeds.

Epic: CRDB-1421
Resolves: #147370

Release note: None

KeithCh commented:
TFTR!

bors=asg0451
